src/java/org/apache/cassandra/service/AbstractWriteResponseHandler.java

```diff
@@ -281,4 +315,102 @@ private final void decrementResponseOrExpired()
             }
         }
     }
+
+    /**
+     * We want to send mutations to as many full replicas as we can, and just as many transient replicas
+     * as we need to meet blockFor.
+     */
+    @VisibleForTesting
+    static SpeculationContext createSpeculationContext(ReplicaList replicas,
```

Comment:
They're currently sorted "naturally", do we want to re-sort them by proximity in this case? Currently `getSortedListByProximity` / `sortByProximity` is used for reads and streaming from what I can see.

Reply:
Previously we sent writes to all replicas and then hinted. There was no need to sort because they were all treated equally. But now we care about health based routing. We want to send writes to all full replicas and then select from a sorted list of transient replicas.